File: sdk/java/NexusLinkClient.java
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;

/**
 * NexusLink Java SDK - 内网穿透平台开放 API 客户端
 *
 * 对接 NexusLink v0.6.0+ 的 /api/v1/* 开放接口（X-API-Key 鉴权）。
 * 纯 JDK 11+（java.net.http），无第三方依赖。
 *
 * 用法：
 *   NexusLinkClient api = new NexusLinkClient("http://127.0.0.1:7001", "dev_api_key_789");
 *   api.createClient("customer-a", "tok_a_1", 3, 1048576);
 *   ClientInfo c = api.getTraffic("customer-a");
 */
public class NexusLinkClient {

    private final String baseUrl;
    private final String apiKey;
    private final HttpClient http;

    public NexusLinkClient(String baseUrl, String apiKey) {
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.apiKey = apiKey;
        this.http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
    }

    /** 通用请求 */
    private String req(String method, String path, String body) throws Exception {
        HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .header("X-API-Key", apiKey)
                .header("Content-Type", "application/json")
                .timeout(Duration.ofSeconds(15));
        if ("POST".equals(method)) {
            b = b.POST(body == null ? HttpRequest.BodyPublishers.noBody() : HttpRequest.BodyPublishers.ofString(body));
        } else if ("PATCH".equals(method)) {
            b = b.method("PATCH", body == null ? HttpRequest.BodyPublishers.noBody() : HttpRequest.BodyPublishers.ofString(body));
        } else if ("DELETE".equals(method)) {
            b = b.DELETE();
        } else {
            b = b.GET();
        }
        HttpResponse<String> resp = http.send(b.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() >= 400) {
            throw new RuntimeException("HTTP " + resp.statusCode() + ": " + resp.body());
        }
        return resp.body();
    }

    // ---------- 客户端管理 ----------

    /** 创建客户端凭据（token 发给客户配置到 client.yaml） */
    public String createClient(String name, String token, int maxTunnels, long maxTrafficBytes) throws Exception {
        return req("POST", "/api/v1/clients",
                String.format("{\"name\":\"%s\",\"token\":\"%s\",\"max_tunnels\":%d,\"max_traffic_bytes\":%d}",
                        name, token, maxTunnels, maxTrafficBytes));
    }

    /** 列出所有客户端 */
    public String listClients() throws Exception {
        return req("GET", "/api/v1/clients", null);
    }

    /** 删除客户端并踢下线 */
    public String deleteClient(String name) throws Exception {
        return req("DELETE", "/api/v1/clients/" + name, null);
    }

    // ---------- 流量查询 ----------

    /** 查询客户端流量 */
    public String getTraffic(String name) throws Exception {
        return req("GET", "/api/v1/clients/" + name + "/traffic", null);
    }

    // ---------- 隧道管理 ----------

    /** 创建隧道（DB 持久化，客户端自动同步注册） */
    public String createProxy(String name, String type, int remotePort, int localPort, String localAddr, String clientName) throws Exception {
        return req("POST", "/api/v1/proxies",
                String.format("{\"name\":\"%s\",\"type\":\"%s\",\"remote_port\":%d,\"local_addr\":\"%s\",\"local_port\":%d,\"client_name\":\"%s\"}",
                        name, type, remotePort, localAddr, localPort, clientName));
    }

    /** 列出所有隧道（DB 定义 + 运行时状态） */
    public String listProxies() throws Exception {
        return req("GET", "/api/v1/proxies", null);
    }

    /** 查看单个隧道详情 */
    public String getProxy(String name) throws Exception {
        return req("GET", "/api/v1/proxies/" + name, null);
    }

    /** 编辑隧道（部分更新，如 {\"remote_port\":8088,\"enabled\":false}） */
    public String updateProxy(String name, String patchJson) throws Exception {
        return req("PATCH", "/api/v1/proxies/" + name, patchJson);
    }

    /** 删除隧道（DB + 运行时下线） */
    public String deleteProxy(String name) throws Exception {
        return req("DELETE", "/api/v1/proxies/" + name, null);
    }

    /** 启用隧道 */
    public String enableProxy(String name) throws Exception {
        return req("POST", "/api/v1/proxies/" + name + "/enable", null);
    }

    /** 停用隧道 */
    public String disableProxy(String name) throws Exception {
        return req("POST", "/api/v1/proxies/" + name + "/disable", null);
    }

    /** 下线隧道（不删除定义） */
    public String closeProxy(String name) throws Exception {
        return req("POST", "/api/v1/proxies/close", "{\"name\":\"" + name + "\"}");
    }

    // ---------- API Key 管理 ----------

    public String listApiKeys() throws Exception {
        return req("GET", "/api/v1/api-keys", null);
    }

    public String createApiKey(String key, String note) throws Exception {
        return req("POST", "/api/v1/api-keys", "{\"key\":\"" + key + "\",\"note\":\"" + note + "\"}");
    }

    public String deleteApiKey(String key) throws Exception {
        return req("DELETE", "/api/v1/api-keys/" + key, null);
    }

    // ==================== 使用示例 ====================
    public static void main(String[] args) throws Exception {
        NexusLinkClient api = new NexusLinkClient("http://127.0.0.1:7001", "dev_api_key_789");

        // 1. 创建客户端
        api.createClient("customer-java", "tok_java_1", 3, 1048576);
        // 2. 列出客户端
        System.out.println("Clients: " + api.listClients());
        // 3. 查询流量
        System.out.println("Traffic: " + api.getTraffic("customer-java"));
        // 4. 清理
        api.deleteClient("customer-java");
    }
}

File: sdk/python/nexuslink_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
NexusLink Python SDK - 内网穿透平台开放 API 客户端

对接 NexusLink v0.6.0+ 的 /api/v1/* 开放接口（X-API-Key 鉴权）。
支持：客户端管理、流量查询、隧道下线、API Key 管理。

依赖：requests（pip install requests）
"""
import requests


class NexusLinkClient:
    def __init__(self, base_url, api_key, timeout=10):
        """
        :param base_url: 服务端地址，如 http://127.0.0.1:7001
        :param api_key:  开放 API 密钥（X-API-Key）
        """
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"X-API-Key": api_key, "Content-Type": "application/json"})

    # ---------- 通用请求 ----------
    def _req(self, method, path, **kwargs):
        resp = self.session.request(method, self.base_url + path, timeout=self.timeout, **kwargs)
        if resp.status_code >= 400:
            raise RuntimeError(f"HTTP {resp.status_code}: {resp.text}")
        return resp.json()

    # ---------- 客户端管理 ----------
    def create_client(self, name, token, max_tunnels=0, max_traffic_bytes=0):
        """创建客户端凭据（token 发给客户配置到 client.yaml）"""
        return self._req("POST", "/api/v1/clients", json={
            "name": name, "token": token,
            "max_tunnels": max_tunnels, "max_traffic_bytes": max_traffic_bytes,
        })

    def list_clients(self):
        """列出所有客户端（含流量/配额/在线状态）"""
        return self._req("GET", "/api/v1/clients").get("clients", [])

    def get_client(self, name):
        """查询单个客户端详情"""
        return self._req("GET", f"/api/v1/clients/{name}").get("client")

    def delete_client(self, name):
        """删除客户端并踢下线"""
        return self._req("DELETE", f"/api/v1/clients/{name}")

    # ---------- 流量查询 ----------
    def get_traffic(self, name):
        """查询客户端流量（bytes_in / bytes_out）"""
        return self._req("GET", f"/api/v1/clients/{name}/traffic").get("client")

    # ---------- 隧道管理 ----------
    def close_proxy(self, name):
        """下线指定隧道"""
        return self._req("POST", "/api/v1/proxies/close", json={"name": name})

    # ---------- API Key 管理 ----------
    def list_api_keys(self):
        return self._req("GET", "/api/v1/api-keys").get("api_keys", [])

    def create_api_key(self, key, note=""):
        return self._req("POST", "/api/v1/api-keys", json={"key": key, "note": note})

    def delete_api_key(self, key):
        return self._req("DELETE", f"/api/v1/api-keys/{key}")


# ==================== 使用示例 ====================
if __name__ == "__main__":
    import sys

    # 用法: nexuslink_client.py [base_url] [api_key] [demo]
    base_url = sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:7001"
    api_key = sys.argv[2] if len(sys.argv) > 2 else "dev_api_key_789"
    api = NexusLinkClient(base_url, api_key)

    if len(sys.argv) > 3 and sys.argv[3] == "demo":
        # 完整对接流程示例
        print("1. 创建客户端")
        api.create_client("customer-demo", "tok_demo_1", max_tunnels=3, max_traffic_bytes=1048576)
        print("2. 列出客户端")
        for c in api.list_clients():
            print(f"   {c['name']}: {c['bytesIn']}B in / {c['bytesOut']}B out, 隧道 {c['proxyCount']}")
        print("3. 查询流量")
        print("   ", api.get_traffic("customer-demo"))
        print("4. 创建 API Key")
        api.create_api_key("key_demo_2", "demo")
        print("5. 清理")
        api.delete_api_key("key_demo_2")
        api.delete_client("customer-demo")
    else:
        # 基础查询
        for c in api.list_clients():
            print(f"{c['name']}\t{c['bytesIn']}B in\t{c['bytesOut']}B out\t隧道 {c['proxyCount']}")
